### PR TITLE
fix: prevent self-reinforcing Blacksmith runner selection loop

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Summary
- `count_queued_jobs` was counting ALL queued jobs regardless of runner type, including jobs already routed to Blacksmith
- This created a feedback loop: once Blacksmith was selected, its queued jobs inflated the backlog count, causing subsequent runs to also select Blacksmith while GitHub-hosted runners sat idle
- Added `targets_github_hosted_runner()` filter to exclude `blacksmith-*` and `self-hosted` labeled jobs from the backlog count

## Test plan
- [ ] Verify `test_select_dynamic_runner.py` passes (6 tests including new `test_count_queued_jobs_excludes_blacksmith_jobs`)
- [ ] Monitor runner usage after merge to confirm GitHub-hosted runners are utilized when backlog is low
- [ ] Verify Blacksmith is still selected when GitHub runner backlog genuinely exceeds threshold

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by thepastaclaw** [comment]
### Review Gate

Commit: `bf81f26b`

- ⏳ **Debounce**: 1m ago (need 30m)
- ✅ **CI checks**: builds passed, 0/0 tests passed
- ✅ **CodeRabbit review**: comment found
- ✅ **Off-peak hours**: off-peak (08:00 PM PT Sunday)

- [ ] Run review now (check to override)

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Clean, well-targeted fix that breaks the self-reinforcing Blacksmith runner selection loop by filtering out non-GitHub-hosted jobs from the backlog count. The new `targets_github_hosted_runner` function correctly excludes `blacksmith-*` and `self-hosted` labels, and the new test directly exercises the exclusion path. No blocking issues found.

*Reviewed commit: 5531691e*

🟡 1 suggestion(s) | 💬 1 nitpick(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `.github/workflows/test_select_dynamic_runner.py`:
- [SUGGESTION] lines 46-77: No test for jobs with empty or missing labels
  The GitHub Actions API can return jobs with an empty `labels` array (e.g., reusable workflow jobs before runner resolution) or theoretically omit the key entirely. `targets_github_hosted_runner` handles this correctly — `job.get("labels", [])` defaults to an empty list, the loop body never executes, and the function returns `True`. However, no test asserts this behavior. Adding a `{"status": "queued", "labels": []}` entry to the existing test data and bumping the expected count would lock in this edge case and prevent future regressions if the function's default-return semantics are accidentally changed.
```
</details>

File: .github/workflows/test_select_dynamic_runner.py [hunk 1]
```diff
@@ -43,6 +43,39 @@ def fetch_json(url):
...
+                    ]
+                },
+                {},
+            ),
+        }
+
+        def fetch_json(url):
+            return responses[url]
+
+        self.assertEqual(MODULE.count_queued_jobs(fetch_json, [repo]), 1)
```
Comment (thepastaclaw) [inline .github/workflows/test_select_dynamic_runner.py:77]: **🟡 Suggestion: No test for jobs with empty or missing labels**

The GitHub Actions API can return jobs with an empty `labels` array (e.g., reusable workflow jobs before runner resolution) or theoretically omit the key entirely. `targets_github_hosted_runner` handles this correctly — `job.get("labels", [])` defaults to an empty list, the loop body never executes, and the function returns `True`. However, no test asserts this behavior. Adding a `{"status": "queued", "labels": []}` entry to the existing test data and bumping the expected count would lock in this edge case and prevent future regressions if the function's default-return semantics are accidentally changed.

<sub>source: ['claude']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `.github/workflows/test_select_dynamic_runner.py`:
- [SUGGESTION] lines 46-77: No test for jobs with empty or missing labels
  The GitHub Actions API can return jobs with an empty `labels` array (e.g., reusable workflow jobs before runner resolution) or theoretically omit the key entirely. `targets_github_hosted_runner` handles this correctly — `job.get("labels", [])` defaults to an empty list, the loop body never executes, and the function returns `True`. However, no test asserts this behavior. Adding a `{"status": "queued", "labels": []}` entry to the existing test data and bumping the expected count would lock in this edge case and prevent future regressions if the function's default-return semantics are accidentally changed.
```
</details>

File: .github/workflows/select_dynamic_runner.py [hunk 2]
```diff
@@ -78,6 +81,19 @@ def iter_pages(
...
+    """Return True if the job targets GitHub-hosted runners, not Blacksmith or self-hosted."""
+    for label in job.get("labels", []):
+        if not isinstance(label, str):
+            continue
+        if label == SELF_HOSTED_LABEL:
+            return False
+        for prefix in NON_GITHUB_HOSTED_RUNNER_PREFIXES:
+            if label.startswith(prefix):
+                return False
+    return True
```
Comment (thepastaclaw) [inline .github/workflows/select_dynamic_runner.py:94]: **💬 Nitpick: Case-sensitive label matching assumes stable casing from GitHub/Blacksmith APIs**

The comparisons at lines 89 and 92 are case-sensitive. GitHub Actions runner labels are case-insensitive for routing but the API returns them in stored casing. Since `self-hosted` is GitHub's standard convention and `blacksmith-*` labels are machine-generated with stable casing, this is extremely unlikely to cause issues in practice. A `.lower()` normalization would add resilience at negligible cost, but this is cosmetic given the controlled label sources.

<details>
<summary>💡 Suggested change</summary>

```suggestion
def targets_github_hosted_runner(job: Dict) -> bool:
    """Return True if the job targets GitHub-hosted runners, not Blacksmith or self-hosted."""
    for label in job.get("labels", []):
        if not isinstance(label, str):
            continue
        lower = label.lower()
        if lower == SELF_HOSTED_LABEL:
            return False
        for prefix in NON_GITHUB_HOSTED_RUNNER_PREFIXES:
            if lower.startswith(prefix):
                return False
    return True
```

</details>

<sub>source: ['claude']</sub>

**Review by UdjinM6 — APPROVED** [review]
utACK 5531691e007bf4365d1ddca5670852269bf9e7d6
